### ISSUE TITLE
Make Axsios pluggable

### ISSUE THREAD

**Issue by hitmands** [axios/axios#1108]
This is a feature request.

Axsios is currently implemented via `Promises` but could be good to **make it pluggable**, 
in this way, for example, we could create a plugin to make it `Observable` (or middleware, or just callback) based.

**Comment by robaxelsen** [comment]
Thanks for your suggestion, @hitmands. Would you be able to submit a PR for this?

**Comment by alinnert** [comment]
Just a thought: I also just had this idea while reading the `Observable` spec. The spec mentions a Symbol that allows an object to be converted into an Observable:

~~~js
const obj = {
  ...otherProps,
  [Symbol.observable] () { return new Observable(...) }
}

Observable.from(obj)
~~~

Plus: RxJS has a method `Observable.fromPromise()`

Now there are two options:

1) Axios (either itself or a plugin) adds this property to every `Promise` returned by axios.
2) Someone adds an ES proposal to add `Symbol.observable` to `Promise.prototype` (as far as I know there's none).

This way an axios result could be converted into an `Observable` by doing:

~~~js
const result = axios.get(...)
Observable.from(result)
~~~

In case of option 2 this is possible with *every* `Promise` and axios would supports this out of the box without any change.

What do you think?

**Comment by hitmands** [comment]
That's it should be made!

**Comment by alinnert** [comment]
Here is my request to implement `Promise.prototype[Symbol.observable]` as part of ECMAscript:

https://github.com/tc39/proposal-observable/issues/181

**Comment by hitmands** [comment]
I think we should provide a basic implementation, just to start a conversation.

**Comment by aze3ma** [comment]
@hitmands that's sounds a cool idea and I would like to contribute also I am wondering if you think to re-implementing a wrapper from scratch or just working with something like `Observable.from()` method.

**Comment by hitmands** [comment]
Hi @aze3ma,
I've been thinking about it and based on others contributions, 
it came out that creating an api which would be cleaner than `Observable.from()` is very difficult.

Axios is completely `promise based` and `rxjs` pairs very well with promises, so, there is probably nothing else to do to make it more rx-friendly.

```javascript
Observable
  .from([1, 2, 3, 4])
  .flatMap(postId => axsios.get('/api/posts/${postId}`))
  .subscribe((post) => {
    console.log(`post ${post.id}`, post);
  })
;
```

**Comment by alinnert** [comment]
@hitmands wow, I didn't know `flatMap` can also return `Promise`s directly. (I tested it, it really works.)

Anyway, Rx is one thing, but it's still open if native Observables will support this as well. It might be necessary to implement `Symbol.Observable` in Axios' result object. (the one that gets returned directly by `axios.get()` etc. directly)

(btw. you sometimes write "axsios" instead of "axios", but I guess it's a typo?)

**Comment by hitmands** [comment]
they may help the discussion: https://itnext.io/working-with-axios-and-rxjs-to-make-simple-ajax-service-module-6fda9ecdaf9f

**Comment by aze3ma** [comment]
@hitmands this is cool I think it would be a helpful for kicking off.

**Comment by aze3ma** [comment]
Hello all does anyone has idea about how we can kick-off the implementation of something like that based purely on observables and if so anyone has already start or something can show us something to start dive into this together.

**Comment by kwonoj** [comment]
Also: I have spinned off https://github.com/kwonoj/oxid for `based purely on observables` related with PR proposals. fyi if anyone's interested.

**Comment by aze3ma** [comment]
@kwonoj that's interesting I am in 🎉

**Comment by chinesedfan** [comment]
Closed in favor of #2412.
